pkg/controller/extension/actuator.go

```diff
@@ -0,0 +1,248 @@
+// SPDX-FileCopyrightText: 2025 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package extension
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	"github.com/gardener/gardener/extensions/pkg/controller/extension"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/gardener/gardener/pkg/extensions"
+	gutil "github.com/gardener/gardener/pkg/utils/gardener"
+	utilsimagevector "github.com/gardener/gardener/pkg/utils/imagevector"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+	"github.com/go-logr/logr"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/gardener/gardener-extension-shoot-cert-service/imagevector"
+	"github.com/gardener/gardener-extension-shoot-cert-service/pkg/apis/config"
+	"github.com/gardener/gardener-extension-shoot-cert-service/pkg/apis/service"
+	"github.com/gardener/gardener-extension-shoot-cert-service/pkg/apis/service/v1alpha1"
+	"github.com/gardener/gardener-extension-shoot-cert-service/pkg/apis/service/validation"
+)
+
+// ActuatorName is the name of the Certificate Service actuator.
+const ActuatorName = "shoot-cert-service-actuator"
+
+// NewActuator returns an actuator responsible for Extension resources.
+func NewActuator(mgr manager.Manager, config config.Configuration, extensionClass extensionsv1alpha1.ExtensionClass) extension.Actuator {
+	return &actuator{
+		client:         mgr.GetClient(),
+		config:         mgr.GetConfig(),
+		decoder:        serializer.NewCodecFactory(mgr.GetScheme(), serializer.EnableStrict).UniversalDecoder(),
+		logger:         log.Log.WithName(ActuatorName),
+		serviceConfig:  config,
+		extensionClass: extensionClass,
+	}
+}
+
+type actuator struct {
+	client         client.Client
+	config         *rest.Config
+	decoder        runtime.Decoder
+	extensionClass extensionsv1alpha1.ExtensionClass
+
+	serviceConfig config.Configuration
+
+	logger logr.Logger
+}
+
+// Reconcile the Extension resource.
+func (a *actuator) Reconcile(ctx context.Context, log logr.Logger, ex *extensionsv1alpha1.Extension) error {
+	var (
+		namespace         = ex.GetNamespace()
+		isShootDeployment = gutil.IsShootNamespace(namespace)
```

Comment:
A little heads up: In the next Gardener release we can use the [GetExtensionClassOrDefault utility](https://github.com/gardener/gardener/blob/df16d9b9a5b4603fe5f6f495bad260fc4f892688/pkg/apis/extensions/v1alpha1/helper/extension.go#L14) to differentiate the deployment cases. Working with the extension class here will probably be the best practice for extensions.